---
title: "test normalizeGeneByImage fun within HER2 dataset"
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
date: 'Compiled: `r Sys.Date()`'
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  time_it = TRUE
)
```


# Overview

```{r install, message = F, warning = F, echo = F}
library(tidyverse)
library(Seurat)
library(stImage)
library(patchwork)
library(Cairo)
library(reticulate)
use_condaenv(condaenv = "myenv", conda = "/home/wangy67/miniconda3/bin/conda")

```


# Prepare input file

```{r define, echo = T, eval = T}
workDir="/data/cqs/ywang/Zhao_image/stImage/SpatialTranscriptomicsImage/stImage/Yuankai/st/"
setwd(workDir)
tstamp <- gsub("-", "", as.character(Sys.Date()))
coordinateFile="/data/cqs/ywang/source/analysis/H1_position.csv"
positionTable=read.csv(coordinateFile, header = F, row.names = 1)
positionTable=positionTable[,c(4,5)]
countTableFile="/data/cqs/ywang/source/analysis/H1.tsv"
sampleName="HER2.H1"
```

```{r, echo = T, eval = F}
rawimg <- "/data/cqs/ywang/source/analysis/HE_BT24044_D2.jpg"

countTable=read.table(countTableFile, header=T, row.names = 1, sep = "\t")
row.names(countTable)=paste0("X",row.names(countTable))

#resnet50
imageFeatureFile="/data/cqs/ywang/Zhao_image/stImage/SpatialTranscriptomicsImage/stImage/Yuankai/st/0_feature_resnet50.csv"
imageFeatures=read.csv(imageFeatureFile,header=TRUE)
row.names(imageFeatures) <- imageFeatures[,1]
imageFeatures <- imageFeatures[, -c(1:4)]
colnames(imageFeatures) <- paste0("ImageFeature", 1:ncol(imageFeatures))

#vgg16
vggimageFeatureFile="/data/cqs/ywang/Zhao_image/stImage/SpatialTranscriptomicsImage/stImage/Yuankai/st/0_feature_vgg16.csv"
vggimageFeatures=read.csv(vggimageFeatureFile,header=TRUE)
row.names(vggimageFeatures) <- vggimageFeatures[,1]
vggimageFeatures <- vggimageFeatures[, -c(1:4)]
colnames(vggimageFeatures) <- paste0("ImageFeature", 1:ncol(vggimageFeatures))


if (identical(sort(row.names(countTable)),sort(row.names(imageFeatures)))) {
  imageFeatures=imageFeatures[row.names(countTable),]
  positionTable=positionTable[row.names(countTable),]
} else {
  stop("Different sample names in countTable and imageFeatures")
}

if (identical(sort(row.names(countTable)),sort(row.names(vggimageFeatures)))) {
  vggimageFeatures=vggimageFeatures[row.names(countTable),]
  positionTable=positionTable[row.names(countTable),]
} else {
  stop("Different sample names in countTable and imageFeatures")
}

```


# Load processed object (processed by previous workflow)

```{r, eval = F}
source("/data/cqs/ywang/source/stImage/R/normalizeGeneByImage.R")
#resnet50
objectRdsName="HER2.H1.Yuankai.resnet50.SeuratObj.rds"
resnetdataObjProcessed <- readRDS(paste0(workDir,basename(objectRdsName), ".processed.rds"))

resnet50_norm_gene_by_image <- normalizeGeneByImage(resnetdataObjProcessed, coordinateNeighborN=6)
#vgg16

vggobjectRdsName="HER2.H1.Yuankai.vgg16.SeuratObj.rds"
vggdataObjProcessed <- readRDS(paste0(workDir,basename(vggobjectRdsName), ".processed.rds"))

vgg16_norm_gene_by_image <- normalizeGeneByImage(vggdataObjProcessed, coordinateNeighborN=6)

#


```

## testing input gene PCs, output normalized gene PCs

```{r, eval = F}
nPCs=30

resnetgenePCs=t(resnetdataObjProcessed@reductions$pca@cell.embeddings[,1:nPCs])
resnet50_norm_genePCs_by_image <- normalizeGeneByImage(resnetdataObjProcessed, coordinateNeighborN=6, geneExp = resnetgenePCs)

vgggenePCs=t(vggdataObjProcessed@reductions$pca@cell.embeddings[,1:nPCs])
vgg16_norm_genePCs_by_image <- normalizeGeneByImage(vggdataObjProcessed, coordinateNeighborN=6, geneExp = vgggenePCs)


```

# Data re-analysis with re-normalized gene exp matrix

## ResNet50

```{r, eval = F}
source("/data/cqs/ywang/source/analysis/MultimodalPreProcess_renorm.R")
objectRdsName <- "HER2.H1.Yuankai.resnet50.SeuratObj.rds"
resnetdataObjProcessed@meta.data <- resnetdataObjProcessed@meta.data[,1:7]
renorm_resnetdataObjProcessed <- MultimodalPreProcess_renorm(resnetdataObjProcessed,
                                                             normalizeMethod = "byImage",
                                                             renormgeneExp = resnet50_norm_gene_by_image,
                                                             geneResolution=1.4,
                                                             imageFeatureResolution = 1.1,
                                                             imagePercentCut=0.1)
renorm_resnetdataObjProcessed <- MultiModalIntegrationVisium(renorm_resnetdataObjProcessed,
                                                normalizeMethod = "SCT",
                                                imageAndGeneResolution=1,
                                                MultimodalMethod = "WNN")
saveRDS(renorm_resnetdataObjProcessed, paste0(workDir,basename(objectRdsName), ".renormgeneExp.processed.rds"))

```


## Vgg-16

```{r, eval = F}
objectRdsName <- "HER2.H1.Yuankai.vgg16.SeuratObj.rds"
renorm_vggdataObjProcessed <- MultimodalPreProcess_renorm(vggdataObjProcessed,
                                                             normalizeMethod = "byImage",
                                                             renormgeneExp = vgg16_norm_gene_by_image,
                                                             geneResolution=1.4,
                                                             imageFeatureResolution = 1.1,
                                                             imagePercentCut=0.1)
renorm_vggdataObjProcessed <- MultiModalIntegrationVisium(renorm_vggdataObjProcessed,
                                                normalizeMethod = "SCT",
                                                imageAndGeneResolution=1,
                                                MultimodalMethod = "WNN")
saveRDS(renorm_vggdataObjProcessed, paste0(workDir,basename(objectRdsName), ".renormgeneExp.processed.rds"))

```


# plot

```{r}
objectRdsName <- "HER2.H1.Yuankai.vgg16.SeuratObj.rds"
renorm_resnetdataObjProcessed <- readRDS(paste0(workDir,basename(objectRdsName), ".renormgeneExp.processed.rds"))
objectRdsName <- "HER2.H1.Yuankai.vgg16.SeuratObj.rds"
renorm_vggdataObjProcessed <- readRDS(paste0(workDir,basename(objectRdsName), ".renormgeneExp.processed.rds"))

```

## DimPlot

```{r, fig.dim = c(12, 10)}

geneClusterName=grep("SCT_snn_res",colnames(renorm_resnetdataObjProcessed@meta.data),value=TRUE)
imageClusterName=grep("ImageFeature_snn_res",colnames(renorm_resnetdataObjProcessed@meta.data),value=TRUE)
integratedClusterName=grep("wsnn_res",colnames(renorm_resnetdataObjProcessed@meta.data),value=TRUE)

p1 <- DimPlot(renorm_resnetdataObjProcessed, reduction = 'st.umap', group.by = geneClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "Gene Expression", sep = "\n"))
p2 <- DimPlot(renorm_resnetdataObjProcessed, reduction = 'if.umap', group.by =imageClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "Image Feature", sep = "\n"))
p3 <- DimPlot(renorm_resnetdataObjProcessed, reduction = 'wnn.umap', group.by =integratedClusterName, label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "WNN", sep = "\n"))

p4 <- DimPlot(renorm_vggdataObjProcessed, reduction = 'st.umap', group.by = geneClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "Gene Expression", sep = "\n"))
p5 <- DimPlot(renorm_vggdataObjProcessed, reduction = 'if.umap', group.by =imageClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "Image Feature", sep = "\n"))
p6 <- DimPlot(renorm_vggdataObjProcessed, reduction = 'wnn.umap', group.by =integratedClusterName, label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "WNN", sep = "\n"))

p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(ncol = 3)

```

### SpatialDimPlot

```{r, fig.dim = c(12, 10)}

DefaultAssay(renorm_resnetdataObjProcessed) <- "Spatial"
p7 <- SpatialDimPlot(renorm_resnetdataObjProcessed, group.by = geneClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "Gene Expression", sep = "\n"))
#+ NoLegend()

DefaultAssay(renorm_resnetdataObjProcessed) <- "ImageFeature"
p8 <- SpatialDimPlot(renorm_resnetdataObjProcessed, group.by = imageClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "Image Feature", sep = "\n"))
#+ NoLegend()

p9 <- SpatialDimPlot(renorm_resnetdataObjProcessed, group.by=integratedClusterName, label = TRUE, label.box = F, repel = TRUE, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "WNN", sep = "\n"))
#+ NoLegend()

DefaultAssay(renorm_vggdataObjProcessed) <- "Spatial"
p10 <- SpatialDimPlot(renorm_vggdataObjProcessed, group.by = geneClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "Gene Expression", sep = "\n"))
#+ NoLegend()

DefaultAssay(renorm_vggdataObjProcessed) <- "ImageFeature"
p11 <- SpatialDimPlot(renorm_vggdataObjProcessed, group.by = imageClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "Image Feature", sep = "\n"))
#+ NoLegend()

p12 <- SpatialDimPlot(renorm_vggdataObjProcessed, group.by=integratedClusterName, label = TRUE, label.box = F, repel = TRUE, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "WNN", sep = "\n"))
#+ NoLegend()

p7 + p8 + p9 + p10 + p11 + p12 + plot_layout(ncol = 3)

```


```{r, eval = F}
CairoPDF("HER2_H1_Yuankai_ResNet50_Vgg16_reNormGeneExp_SpatialDimPlots.pdf", width = 12, height = 10, onefile = T)
p7 + p8 + p9 + p10 + p11 + p12 + plot_layout(ncol = 3)
dev.off()
```

### Weight of SCT and ImageFeature

#### ResNet50

```{r, fig.dim = c(8, 5)}

p13 <- VlnPlot(renorm_resnetdataObjProcessed, features = "SCT.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()
p14 <- VlnPlot(renorm_resnetdataObjProcessed, features = "ImageFeature.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()

p13 + p14
```

#### Vgg16

```{r, fig.dim = c(8, 5)}

p15 <- VlnPlot(renorm_vggdataObjProcessed, features = "SCT.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()
p16 <- VlnPlot(renorm_vggdataObjProcessed, features = "ImageFeature.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()

p15 + p16
```


### ARI with pathologist review region: ResNet50

```{r}
library(clues)
#pathologist review region
reviewRegions=read.delim("/scratch/cqs/zhaos/Spatial/SpatialTranscriptomicsImage/her2st/data/ST-pat/lbl/H1_labeled_coordinates.tsv")
temp1=order(renorm_resnetdataObjProcessed@images$images@coordinates$V5,renorm_resnetdataObjProcessed@images$images@coordinates$V6)
#head(positionTable[temp1,])
temp2=order(reviewRegions$y,reviewRegions$x)
#head(reviewRegions[temp2,])
renorm_resnetdataObjProcessed@meta.data$TrueLabel=NA
renorm_resnetdataObjProcessed@meta.data$TrueLabel[temp1]=reviewRegions$label[temp2]
#visulization of review regions
SpatialDimPlot(renorm_resnetdataObjProcessed, group.by = "TrueLabel", label = TRUE, label.box = F, repel = TRUE, label.size = 2,pt.size.factor = 10) 
p6
predictedCluster=as.character(renorm_resnetdataObjProcessed@meta.data$"SCT_snn_res.1.4")
#table(renorm_resnetdataObjProcessed@meta.data$"wsnn_res.1",renorm_resnetdataObjProcessed@meta.data$TrueLabel)
table(predictedCluster,renorm_resnetdataObjProcessed@meta.data$TrueLabel)
predictedCluster[predictedCluster==0]="connective tissue"
predictedCluster[predictedCluster==1]="invasive cancer"
predictedCluster[predictedCluster==2]="cancer in situ"
predictedCluster[predictedCluster==3]="adipose tissue"
predictedCluster[predictedCluster==4]="adipose tissue"
predictedCluster[predictedCluster==5]="breast glands"
predictedCluster[predictedCluster==6]="immune infiltrate"
trueLabel=renorm_resnetdataObjProcessed@meta.data$TrueLabel
naInd=which(trueLabel=="undetermined")
    adjustedRand(
      as.integer(as.factor(predictedCluster[-naInd])),
      as.integer(as.factor(trueLabel[-naInd])),
      randMethod="HA"
    )
#0.3367327
```


```{r, eval = F}
    
library(nnet)
library(DescTools)
PseudoR2Result=NULL 
for (i in c(1:20)) {
  #fit mutinorm regression model to get PseudoR2
temp=cbind(renorm_resnetdataObjProcessed@reductions$ipca@cell.embeddings[-naInd,1:i],
                                                  renorm_resnetdataObjProcessed@reductions$pca@cell.embeddings[-naInd,1:i])
modelResult <- multinom(trueLabel[-naInd] ~ temp, maxit = 1000, MaxNWts = 2000,model = TRUE)
PseudoR2Result=c(PseudoR2Result,PseudoR2(modelResult,c("McFaddenAdj")))
}
names(PseudoR2Result)=1:20
plot(PseudoR2Result,type="b")
max(PseudoR2Result)
#0.7746901
#marker gene expression
selectedGenes=c("DDX54", "TRAF2")
VlnPlot(renorm_resnetdataObjProcessed, features = selectedGenes,assay = "SCT")
DotPlot(renorm_resnetdataObjProcessed, features = selectedGenes,assay = "SCT")
		
```


### ARI with pathologist review region: Vgg16


```{r, warning = F, echo = T}
reviewRegions=read.delim("/scratch/cqs/zhaos/Spatial/SpatialTranscriptomicsImage/her2st/data/ST-pat/lbl/H1_labeled_coordinates.tsv")
temp3=order(renorm_vggdataObjProcessed@images$images@coordinates$V5,renorm_vggdataObjProcessed@images$images@coordinates$V6)
#head(positionTable[temp1,])
temp4=order(reviewRegions$y,reviewRegions$x)
#head(reviewRegions[temp2,])
renorm_vggdataObjProcessed@meta.data$TrueLabel=NA
renorm_vggdataObjProcessed@meta.data$TrueLabel[temp3]=reviewRegions$label[temp4]

vggpredictedCluster=as.character(renorm_vggdataObjProcessed@meta.data$"SCT_snn_res.1.4")
#table(renorm_vggdataObjProcessed@meta.data$"wsnn_res.1",renorm_vggdataObjProcessed@meta.data$TrueLabel)
table(vggpredictedCluster,renorm_vggdataObjProcessed@meta.data$TrueLabel)
vggpredictedCluster[vggpredictedCluster==0]="connective tissue"
vggpredictedCluster[vggpredictedCluster==1]="invasive cancer"
vggpredictedCluster[vggpredictedCluster==2]="adipose tissue"
vggpredictedCluster[vggpredictedCluster==3]="cancer in situ"
vggpredictedCluster[vggpredictedCluster==4]="immune infiltrate"
vggpredictedCluster[vggpredictedCluster==5]="breast glands"
vggtrueLabel=renorm_vggdataObjProcessed@meta.data$TrueLabel
naInd=which(vggtrueLabel=="undetermined")
    adjustedRand(
      as.integer(as.factor(vggpredictedCluster[-naInd])),
      as.integer(as.factor(vggtrueLabel[-naInd])),
      randMethod="HA"
    )
#0.6004867


```


```{r, eval = F}
    
library(nnet)
library(DescTools)
PseudoR2Result=NULL 
for (i in c(1:20)) {
  #fit mutinorm regression model to get PseudoR2
temp=cbind(renorm_vggdataObjProcessed@reductions$ipca@cell.embeddings[-naInd,1:i],
                                                  renorm_vggdataObjProcessed@reductions$pca@cell.embeddings[-naInd,1:i])
modelResult <- multinom(trueLabel[-naInd] ~ temp, maxit = 1000, MaxNWts = 2000,model = TRUE)
PseudoR2Result=c(PseudoR2Result,PseudoR2(modelResult,c("McFaddenAdj")))
}
names(PseudoR2Result)=1:20
plot(PseudoR2Result,type="b")
max(PseudoR2Result)
#0.750693
```


### marker gene expression

```{r}
selectedGenes=c("DDX54", "TRAF2")
DefaultAssay(renorm_vggdataObjProcessed) <- "Spatial"
SpatialFeaturePlot(renorm_vggdataObjProcessed, pt.size.factor = 8, features = selectedGenes, slot = "data")

VlnPlot(renorm_vggdataObjProcessed, features = selectedGenes, assay = "Spatial")
#DotPlot(vggdataObjProcessed, features = selectedGenes,assay = "SCT")
		
```




# Data re-analysis with re-normalized gene exp PC matrix

## ResNet50

```{r, eval = F}
source("/data/cqs/ywang/source/analysis/MultimodalPreProcess_renorm.R")

objectRdsName <- "HER2.H1.Yuankai.resnet50.SeuratObj.rds"
renormPCs_resnetdataObjProcessed <- MultimodalPreProcess_renorm(resnetdataObjProcessed,
                                                             normalizeMethod = "SCT",
                                                             GeneFeaturePCs = t(resnet50_norm_genePCs_by_image),
                                                             geneResolution=1.4,
                                                             imageFeatureResolution = 1.1,
                                                             imagePercentCut=0.1)
renormPCs_resnetdataObjProcessed <- MultiModalIntegrationVisium(renormPCs_resnetdataObjProcessed,
                                                normalizeMethod = "SCT",
                                                imageAndGeneResolution=1,
                                                MultimodalMethod = "WNN")
saveRDS(renormPCs_resnetdataObjProcessed, paste0(workDir,basename(objectRdsName), ".renormgeneExpPCs.processed.rds"))

```


## Vgg-16

```{r, eval = F}
objectRdsName <- "HER2.H1.Yuankai.vgg16.SeuratObj.rds"
renormPCs_vggdataObjProcessed <- MultimodalPreProcess_renorm(vggdataObjProcessed,
                                                             normalizeMethod = "SCT",
                                                             GeneFeaturePCs = t(vgg16_norm_genePCs_by_image),
                                                             geneResolution=1.4,
                                                             imageFeatureResolution = 1.1,
                                                             imagePercentCut=0.1)
renormPCs_vggdataObjProcessed <- MultiModalIntegrationVisium(renormPCs_vggdataObjProcessed,
                                                normalizeMethod = "SCT",
                                                imageAndGeneResolution=1,
                                                MultimodalMethod = "WNN")
saveRDS(renormPCs_vggdataObjProcessed, paste0(workDir,basename(objectRdsName), ".renormgeneExpPCs.processed.rds"))

```


# plot


```{r}
objectRdsName <- "HER2.H1.Yuankai.resnet50.SeuratObj.rds"
renormPCs_resnetdataObjProcessed <- readRDS(paste0(workDir,basename(objectRdsName), ".renormgeneExpPCs.processed.rds"))
objectRdsName <- "HER2.H1.Yuankai.vgg16.SeuratObj.rds"
renormPCs_vggdataObjProcessed <- readRDS(paste0(workDir,basename(objectRdsName), ".renormgeneExpPCs.processed.rds"))

```


## DimPlot

```{r, fig.dim = c(12, 10)}

geneClusterName=grep("SCT_snn_res",colnames(renormPCs_resnetdataObjProcessed@meta.data),value=TRUE)
imageClusterName=grep("ImageFeature_snn_res",colnames(renormPCs_resnetdataObjProcessed@meta.data),value=TRUE)
integratedClusterName=grep("wsnn_res",colnames(renormPCs_resnetdataObjProcessed@meta.data),value=TRUE)

p1 <- DimPlot(renormPCs_resnetdataObjProcessed, reduction = 'st.umap', group.by = geneClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "Gene Expression", sep = "\n"))
p2 <- DimPlot(renormPCs_resnetdataObjProcessed, reduction = 'if.umap', group.by =imageClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "Image Feature", sep = "\n"))
p3 <- DimPlot(renormPCs_resnetdataObjProcessed, reduction = 'wnn.umap', group.by =integratedClusterName, label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_ResNet50", "WNN", sep = "\n"))

p4 <- DimPlot(renormPCs_vggdataObjProcessed, reduction = 'st.umap', group.by = geneClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "Gene Expression", sep = "\n"))
p5 <- DimPlot(renormPCs_vggdataObjProcessed, reduction = 'if.umap', group.by =imageClusterName, label = TRUE, 
              repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "Image Feature", sep = "\n"))
p6 <- DimPlot(renormPCs_vggdataObjProcessed, reduction = 'wnn.umap', group.by =integratedClusterName, label = TRUE, repel = TRUE, label.size = 2.5) + NoLegend() + ggtitle(paste("YK_Vgg16", "WNN", sep = "\n"))

p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(ncol = 3)

```

### SpatialDimPlot

```{r, fig.dim = c(12, 10)}

DefaultAssay(renormPCs_resnetdataObjProcessed) <- "Spatial"
p7 <- SpatialDimPlot(renormPCs_resnetdataObjProcessed, group.by = geneClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "Gene Expression", sep = "\n"))
#+ NoLegend()

DefaultAssay(renormPCs_resnetdataObjProcessed) <- "ImageFeature"
p8 <- SpatialDimPlot(renormPCs_resnetdataObjProcessed, group.by = imageClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "Image Feature", sep = "\n"))
#+ NoLegend()

p9 <- SpatialDimPlot(renormPCs_resnetdataObjProcessed, group.by=integratedClusterName, label = TRUE, label.box = F, repel = TRUE, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_ResNet50", "WNN", sep = "\n"))
#+ NoLegend()

DefaultAssay(renormPCs_vggdataObjProcessed) <- "Spatial"
p10 <- SpatialDimPlot(renormPCs_vggdataObjProcessed, group.by = geneClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "Gene Expression", sep = "\n"))
#+ NoLegend()

DefaultAssay(renormPCs_vggdataObjProcessed) <- "ImageFeature"
p11 <- SpatialDimPlot(renormPCs_vggdataObjProcessed, group.by = imageClusterName, label = TRUE, label.box = F, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "Image Feature", sep = "\n"))
#+ NoLegend()

p12 <- SpatialDimPlot(renormPCs_vggdataObjProcessed, group.by=integratedClusterName, label = TRUE, label.box = F, repel = TRUE, label.size = 2, pt.size.factor = 10) + ggtitle(paste("YK_Vgg16", "WNN", sep = "\n"))
#+ NoLegend()

p7 + p8 + p9 + p10 + p11 + p12 + plot_layout(ncol = 3)

```


```{r, eval = F}
CairoPDF("HER2_H1_Yuankai_ResNet50_Vgg16_reNormGeneExpPCs_SpatialDimPlots.pdf", width = 12, height = 10, onefile = T)
p7 + p8 + p9 + p10 + p11 + p12 + plot_layout(ncol = 3)
dev.off()
```

### Weight of SCT and ImageFeature

#### ResNet50

```{r, fig.dim = c(8, 5)}

p13 <- VlnPlot(renormPCs_resnetdataObjProcessed, features = "SCT.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()
p14 <- VlnPlot(renormPCs_resnetdataObjProcessed, features = "ImageFeature.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()

p13 + p14
```

#### Vgg16

```{r, fig.dim = c(8, 5)}

p15 <- VlnPlot(renormPCs_vggdataObjProcessed, features = "SCT.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()
p16 <- VlnPlot(renormPCs_vggdataObjProcessed, features = "ImageFeature.weight", group.by = integratedClusterName, sort = TRUE, pt.size = 0.1) +
  NoLegend()

p15 + p16
```


### ARI with pathologist review region: ResNet50

```{r}
library(clues)
#pathologist review region
reviewRegions=read.delim("/scratch/cqs/zhaos/Spatial/SpatialTranscriptomicsImage/her2st/data/ST-pat/lbl/H1_labeled_coordinates.tsv")
temp1=order(renormPCs_resnetdataObjProcessed@images$images@coordinates$V5,renormPCs_resnetdataObjProcessed@images$images@coordinates$V6)
#head(positionTable[temp1,])
temp2=order(reviewRegions$y,reviewRegions$x)
#head(reviewRegions[temp2,])
renormPCs_resnetdataObjProcessed@meta.data$TrueLabel=NA
renormPCs_resnetdataObjProcessed@meta.data$TrueLabel[temp1]=reviewRegions$label[temp2]
#visulization of review regions
SpatialDimPlot(renormPCs_resnetdataObjProcessed, group.by = "TrueLabel", label = TRUE, label.box = F, repel = TRUE, label.size = 2,pt.size.factor = 10) 
p6
predictedCluster=as.character(renormPCs_resnetdataObjProcessed@meta.data$"wsnn_res.1")
#table(renormPCs_resnetdataObjProcessed@meta.data$"wsnn_res.1",renormPCs_resnetdataObjProcessed@meta.data$TrueLabel)
table(predictedCluster,renormPCs_resnetdataObjProcessed@meta.data$TrueLabel)
predictedCluster[predictedCluster==0]="connective tissue"
predictedCluster[predictedCluster==1]="cancer in situ"
predictedCluster[predictedCluster==2]="invasive cancer"
predictedCluster[predictedCluster==3]="adipose tissue"
predictedCluster[predictedCluster==4]="adipose tissue"
predictedCluster[predictedCluster==5]="immune infiltrate"
predictedCluster[predictedCluster==6]="breast glands"
trueLabel=renormPCs_resnetdataObjProcessed@meta.data$TrueLabel
naInd=which(trueLabel=="undetermined")
    adjustedRand(
      as.integer(as.factor(predictedCluster[-naInd])),
      as.integer(as.factor(trueLabel[-naInd])),
      randMethod="HA"
    )
#0.5449543
```


```{r, eval = F}
    
library(nnet)
library(DescTools)
PseudoR2Result=NULL 
for (i in c(1:20)) {
  #fit mutinorm regression model to get PseudoR2
temp=cbind(renormPCs_resnetdataObjProcessed@reductions$ipca@cell.embeddings[-naInd,1:i],
                                                  renormPCs_resnetdataObjProcessed@reductions$pca@cell.embeddings[-naInd,1:i])
modelResult <- multinom(trueLabel[-naInd] ~ temp, maxit = 1000, MaxNWts = 2000,model = TRUE)
PseudoR2Result=c(PseudoR2Result,PseudoR2(modelResult,c("McFaddenAdj")))
}
names(PseudoR2Result)=1:20
plot(PseudoR2Result,type="b")
max(PseudoR2Result)
#0.7746901
#marker gene expression
selectedGenes=c("DDX54", "TRAF2")
VlnPlot(renormPCs_resnetdataObjProcessed, features = selectedGenes,assay = "SCT")
DotPlot(renormPCs_resnetdataObjProcessed, features = selectedGenes,assay = "SCT")
		
```


### ARI with pathologist review region: Vgg16


```{r, warning = F, echo = T}
reviewRegions=read.delim("/scratch/cqs/zhaos/Spatial/SpatialTranscriptomicsImage/her2st/data/ST-pat/lbl/H1_labeled_coordinates.tsv")
temp3=order(renormPCs_vggdataObjProcessed@images$images@coordinates$V5,renormPCs_vggdataObjProcessed@images$images@coordinates$V6)
#head(positionTable[temp1,])
temp4=order(reviewRegions$y,reviewRegions$x)
#head(reviewRegions[temp2,])
renormPCs_vggdataObjProcessed@meta.data$TrueLabel=NA
renormPCs_vggdataObjProcessed@meta.data$TrueLabel[temp3]=reviewRegions$label[temp4]

vggpredictedCluster=as.character(renormPCs_vggdataObjProcessed@meta.data$"wsnn_res.1")
#table(renormPCs_vggdataObjProcessed@meta.data$"wsnn_res.1",renormPCs_vggdataObjProcessed@meta.data$TrueLabel)
table(vggpredictedCluster,renormPCs_vggdataObjProcessed@meta.data$TrueLabel)
vggpredictedCluster[vggpredictedCluster==0]="connective tissue"
vggpredictedCluster[vggpredictedCluster==1]="invasive cancer"
vggpredictedCluster[vggpredictedCluster==2]="adipose tissue"
vggpredictedCluster[vggpredictedCluster==3]="cancer in situ"
vggpredictedCluster[vggpredictedCluster==4]="immune infiltrate"
vggpredictedCluster[vggpredictedCluster==5]="breast glands"
vggtrueLabel=renormPCs_vggdataObjProcessed@meta.data$TrueLabel
naInd=which(vggtrueLabel=="undetermined")
    adjustedRand(
      as.integer(as.factor(vggpredictedCluster[-naInd])),
      as.integer(as.factor(vggtrueLabel[-naInd])),
      randMethod="HA"
    )
#0.5750029 


```


```{r, eval = F}
    
library(nnet)
library(DescTools)
PseudoR2Result=NULL 
for (i in c(1:20)) {
  #fit mutinorm regression model to get PseudoR2
temp=cbind(renormPCs_vggdataObjProcessed@reductions$ipca@cell.embeddings[-naInd,1:i],
                                                  renormPCs_vggdataObjProcessed@reductions$pca@cell.embeddings[-naInd,1:i])
modelResult <- multinom(trueLabel[-naInd] ~ temp, maxit = 1000, MaxNWts = 2000,model = TRUE)
PseudoR2Result=c(PseudoR2Result,PseudoR2(modelResult,c("McFaddenAdj")))
}
names(PseudoR2Result)=1:20
plot(PseudoR2Result,type="b")
max(PseudoR2Result)
#0.750693
```


### marker gene expression

```{r}
selectedGenes=c("DDX54", "TRAF2")
DefaultAssay(renormPCs_vggdataObjProcessed) <- "Spatial"
SpatialFeaturePlot(renormPCs_vggdataObjProcessed, pt.size.factor = 8, features = selectedGenes, slot = "data")

VlnPlot(renormPCs_vggdataObjProcessed, features = selectedGenes, assay = "Spatial")
#DotPlot(vggdataObjProcessed, features = selectedGenes,assay = "SCT")
		
```

