---
title: "Adult_Mouse_Olfactory_Bulb Visium version"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Adult_Mouse_Olfactory_Bulb Visium version}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
date: 'Compiled: `r Sys.Date()`'
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  time_it = TRUE
)
```


# Overview

```{r install, message = F, warning = F, echo = F, eval = T}
library(tidyverse)
library(Seurat)
library(stImage)
library(SpatialPCA)
library(BayesSpace)
library(patchwork)
library(reticulate)
library(tensorflow)
library(keras)
```


# Download raw dataset from 10xgenomics

Adult Mouse Olfactory Bulb (Spatial Gene Expression Dataset by Space Ranger 2.0.0) dataset can be accessed at https://www.10xgenomics.com/resources/datasets/adult-mouse-olfactory-bulb-1-standard-1. 

The following files will be needed:

- [The hdf5 file of the gene expression matrix](https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_filtered_feature_bc_matrix.h5)

- [Spatial imaging data](https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_spatial.tar.gz)

- [raw H&E image (TIFF)](https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_image.tif) (The raw image file is optional. It's only needed if planning to extract image features from the raw image.)


After downloading the above files, supposed that all of the files are put under a folder called "MOB". The spatial image data should be uncompressed and renamed to "spatial" by: 

`tar zxvf Visium_Mouse_Olfactory_Bulb_spatial.tar.gz; mv Visium_Mouse_Olfactory_Bulb_spatial spatial`.



```{r define, echo = T, eval = T}
data_path = "/data/cqs/ywang/Zhao_image/stImage/Adult_Mouse_Olfactory_Bulb"
spatial_path <- paste(data_path, "spatial", sep = "/")
h5_file <- "Visium_Mouse_Olfactory_Bulb_filtered_feature_bc_matrix.h5"

# imgFile <- paste(spatial_path, "tissue_hires_image.png", sep = "/")
rawimg <- "Visium_Mouse_Olfactory_Bulb_image.tif" #optional

#positionTableFile is 10X format, Columns: Tissue; Y,X,ImageY,ImageX
positionTableFile=paste(spatial_path, "tissue_positions_list.csv", sep = "/")

```


<br>

# Extract features

Image features and RGB quantile matrix will be extracted with VGG16 model of ImageNet from the `tissue_hires_image.png` under `spatial` folder.

```{r extract.feature, echo = T, eval = T}
imageFeatures_list <- ExtractFeaturesVisium(spatialDir = spatial_path,
                                              rawImage = T,
                                              imgFile = paste0(data_path, "/", rawimg))

imageFeatures <- imageFeatures_list[["ImageFeature"]]
RGBquantile <- imageFeatures_list[["RGBquantile"]]

#imageFeatures_list2 <- ExtractFeaturesVisium(spatialDir = spatial_path)
```

<br>

# Prepare object containing gene expression data and image feature matrix 

```{r load.data, echo = T, eval = T}
MOB_obj <- LoadImageFeatureVisium(data_path, 
                                  filename = h5_file,
                                  imageFeatures = imageFeatures,
                                  RGBquantile = RGBquantile)

```


<br>

# Preprocess of Spatial matrix and ImageFeature matrix

```{r data.process, message = F, warning = F, echo = T, eval = T, results = 'hide'}
clusterNum <- 8

MOB_obj <- PreProcessing(
  MOB_obj, 
  normalizeMethod = "SCT", 
  pcaDim_s=30, 
  pcaDim_i=30, 
  pcaDim_c = 6,
  DimReducMethod = "PCA",
  prefiltergenePercentCut=0,
  prefilterimagePercentCut=0,
  genePercentCut = 0.2,
  imagePercentCut = 0.2)

MOB_obj <- FindSNNClusters(
  MOB_obj,
  dims = 1:30,
  reduction = "SCTPCA",
  nCluster = clusterNum,
  resolutionMax = 3
)

MOB_obj <- FindSNNClusters(
  MOB_obj,
  dims = 1:30,
  reduction = "ImageFeaturePCA",
  nCluster = clusterNum,
  resolutionMax = 3
)

MOB_obj <- FindSNNClusters(
  MOB_obj,
  dims = 1:6,
  reduction = "RGBPCA",
  nCluster = clusterNum,
  resolutionMax = 3
)

MOB_obj <- MultiModalIntegration(
  MOB_obj,
  MultimodalMethod = "WNN",
  pcaDim_s = 30,
  pcaDim_i = 30,
  reduction.list = list("SCTPCA", "ImageFeaturePCA"),
  nCluster = clusterNum,
  resolutionMax = 3
)

```


# Visualization

## DimPlot

```{r DimPlot, fig.dim = c(12, 4), echo = T, eval = T}
p1 <- DimPlot(MOB_obj, reduction = 'SCTPCAUMAP', group.by = 'seurat_clusters', label = TRUE, 
              repel = TRUE, label.size = 2.5) + ggtitle("SCT") + NoLegend()
p2 <- DimPlot(MOB_obj, reduction = 'ImageFeaturePCAUMAP', group.by = 'seurat_clusters', label = TRUE, 
              repel = TRUE, label.size = 2.5) + ggtitle("ImageFeature") + NoLegend()
p3 <- DimPlot(MOB_obj, reduction = 'weighted.nnUMAP', label = TRUE, repel = TRUE, label.size = 2.5) + ggtitle("WNN") + NoLegend()

p1 + p2 + p3
```

## SpatialDimPlot

```{r SpatialDimPlot, fig.dim = c(12, 4), echo = T, eval = T}
DefaultAssay(MOB_obj) <- "Spatial"
p4 <- SpatialDimPlot(MOB_obj, group.by = "SCTPCA_cluster8", label = TRUE, label.box = F, label.size = 2) + ggtitle("SCT") & NoLegend()

DefaultAssay(MOB_obj) <- "ImageFeature"
p5 <- SpatialDimPlot(MOB_obj, group.by = "ImageFeaturePCA_cluster8", label = TRUE, label.box = F, label.size = 2) + ggtitle("ImageFeature") & NoLegend()

p6 <- SpatialDimPlot(MOB_obj, group.by = "wnn_cluster8", label = TRUE, label.box = F, repel = TRUE, label.size = 2) + ggtitle("WNN") & NoLegend()

p4 + p5 + p6
```


<details>

<summary>**Session Info**</summary>

```{r}
sessionInfo()
```
</details>
