#' Load Visium data
#'
#' @param dataDir a path/folder which contains the dataset generated by
#' SpaceRanger
#' @param filename Directory containing the H5 file specified by \code{filename}
#' and the image data in a subdirectory called \code{spatial}
#' @param imageFeatures image feature matrix extracted from
#' ExtractImageFeature.R or provided by user
#' @param RGBquantile RGB quantile matrix extracted from ExtractImageFeature.R
#' or provided by user
#' @param assay assay name for storing gene expression matrix in seurat object,
#' default value: "Spatial"
#' @param slice Name for the stored image of the tissue slice, default value:
#' "slice1"
#' @param filter.matrix Only keep spots that have been determined to be over
#' tissue, default value: TRUE
#' @param to.upper Converts all feature names to upper case. This can provide
#' an approximate conversion of mouse to human gene names which can be useful
#' in an explorative analysis. For cross-species comparisons, orthologous genes
#' should be identified across species and used instead.
#' @param image An object of class VisiumV1. Typically, an output from
#' \code{\link{Read10X_Image}}
#' @param ... Arguments passed to \code{\link{Read10X_h5}}
#'
#' @return A \code{Seurat} object
#' @importFrom png readPNG
#' @importFrom grid rasterGrob
#' @importFrom jsonlite fromJSON
#' @importFrom Seurat CreateAssayObject Load10X_Spatial
#' @export
#' @concept preprocessing
#'
#' @examples
#' \dontrun{
#' data_dir <- 'path/to/data/directory'
#' list.files(data_dir) # Should show filtered_feature_bc_matrix.h5
#' imageFeatures <- read.csv(ImageFeatureFile)
#' LoadImageFeatureVisium(data.dir = data_dir, filename = h5_file,
#' imageFeatures = imageFeatures)
#' }
#'
LoadImageFeatureVisium <- function(dataDir,
                                   filename = "filtered_feature_bc_matrix.h5",
                                   imageFeatures,
                                   RGBquantile = NULL,
                                   assay = "Spatial",
                                   slice = "slice1",
                                   filter.matrix = TRUE,
                                   to.upper = FALSE,
                                   image = NULL, ...) {
  options(Seurat.object.assay.version = "v3")
  object <- Load10X_Spatial(dataDir,
                            filename = filename,
                            assay = assay,
                            slice = slice,
                            filter.matrix = filter.matrix,
                            to.upper = to.upper,
                            image = image)
  #removing constant features
  constantColumns <- which(apply(imageFeatures, 2,
                                 function(x) sd(x, na.rm=TRUE)) == 0)
  if (length(constantColumns) > 0) {
    imageFeatures <- imageFeatures[, -constantColumns]
    warning(paste0(length(constantColumns),
                   " image features have SD=0 and were removed"))
  }
  imageFeaturesObj <- CreateAssayObject(counts = t(imageFeatures))
  object@assays$ImageFeature <- imageFeaturesObj
  if(!is.null(RGBquantile)) {
    rgbObj <- CreateAssayObject(counts = t(RGBquantile))
    object@assays$RGB <- rgbObj
    message("Image Feature, expression matrix and RGB quantile matrix
            has been successfully loaded!")
  } else {
    message("Image Feature and expression matrix has been successfully loaded!")
  }
  return(object)
}

#' Load data from raw matrix files
#'
#' @param countTable gene (row) by cell (column) matrix of spatial
#' transcriptomic data generated by any platform
#' @param imageFeatures image feature matrix extracted by
#' \code{ExtractFeatures} or provided by user
#' @param positionTable matrix containing a table with rows that correspond to
#' spots, see 'spatial/tissue_positions.csv' in Visium dataset for example
#' @param RGBquantile RGB quantile matrix extracted from ExtractImageFeature.R
#' or provided by user
#' @param project project name when create seurat object
#' @param Class the platform types of 'images' slot, i.e., ST, or Slide-seq
#' @param ... Arguments passed to \code{\link{Read10X_h5}}
#'
#' @return A \code{Seurat} object
#' @importFrom png readPNG
#' @importFrom grid rasterGrob
#' @importFrom jsonlite fromJSON
#' @importFrom Seurat CreateAssayObject CreateSeuratObject
#' @export
#' @concept preprocessing
#'
#' @examples
#' \dontrun{
#' countTable <- read.table(GeneByCellFile)
#' imageFeatures <- read.csv(ImageFeatureFile)
#' positionTable <- read.table(PositionTableFile)
#' LoadImageFeature(countTable = countTable, imageFeatures = imageFeatures,
#' positionTable = positionTable, project = "test")
#' }
#'
LoadImageFeature <- function(countTable,
                             imageFeatures,
                             positionTable,
                             RGBquantile = NULL,
                             project = NULL,
                             Class = "SlideSeq", ...) {
  options(Seurat.object.assay.version = "v3")
  imageFeaturesObj <- CreateAssayObject(counts = t(imageFeatures))
  object <- CreateSeuratObject(countTable, project = project, assay = "Spatial")
  object@assays$ImageFeature <- imageFeaturesObj
  object@images$image <- new(Class = Class,
                             assay = "Spatial",
                             key = "image_",
                             coordinates = positionTable)
  if(!is.null(RGBquantile)) {
    rgbObj <- CreateAssayObject(counts = t(RGBquantile))
    object@assays$RGB <- rgbObj
    message("Image Feature, expression matrix and RGB quantile matrix
            has been successfully loaded!")
  } else {
    message("Image Feature and expression matrix has been successfully loaded!")
  }
  return(object)
}
